pkg/object/mqttproxy/session_manager.go

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"sync"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/object/function/storage"
+)
+
+type (
+	// SessionManager manage the status of session for clients
+	SessionManager struct {
+		broker  *Broker
+		smap    sync.Map
+		store   storage.Storage
+		storeCh chan SessionStore
+		done    chan struct{}
+	}
+
+	// SessionStore for session store, key is session clientID, value is session yaml marshal value
+	SessionStore struct {
+		key   string
+		value string
+	}
+)
+
+func newSessionManager(b *Broker, store storage.Storage) *SessionManager {
+	sm := &SessionManager{
+		broker:  b,
+		store:   store,
+		storeCh: make(chan SessionStore),
```

Comment:
propose to use a buffered channel and not to create separated goroutines in `Session.store()`

Reply:
If use buffered channel, then how big of the channel will be a problem. If Session.store() not use goroutine, the session subscribe and unsubscribe will hold the lock until the store finished which decided on the buffered channel. I think current approach is a little bit more general